examples/audiopanning/main.go

```diff
@@ -0,0 +1,232 @@
+
+	g.player, err = audio.NewPlayer(audioContext, s)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Play the infinite-length stream. This never ends.
+	g.player.Play()
+}
+
+func (g *Game) Update(screen *ebiten.Image) error {
+	g.initAudio()
+	g.r += ((1.0 / 60.0) * 2 * math.Pi) * 0.1 // full cycle every 10 seconds
+	g.xpos = (float64(screenWidth) / 2) + math.Cos(g.r)*(float64(screenWidth)/2)
+	g.panning = lerp(-1, 1, g.xpos/float64(screenWidth))
+	g.panstream.SetPan(g.panning)
+	return nil
+}
+
+func (g *Game) Draw(screen *ebiten.Image) {
+	pos := g.player.Current()
+	if pos > 5*time.Second {
+		pos = (g.player.Current()-5*time.Second)%(4*time.Second) + 5*time.Second
+	}
+	msg := fmt.Sprintf(`TPS: %0.2f
+This is an example using
+stereo audio panning.
+Intro:   0[s] - %[2]d[s]
+Loop:    %[2]d[s] - %[3]d[s]
+Current: %0.2[4]f[s]
+Panning: %.2f`, ebiten.CurrentTPS(), introLengthInSecond, introLengthInSecond+loopLengthInSecond, float64(pos)/float64(time.Second), g.panning)
+	ebitenutil.DebugPrint(screen, msg)
+
+	// draw image to show where the sound is at related to the screen
+	op := &ebiten.DrawImageOptions{}
+	op.GeoM.Translate(g.xpos-float64(img.Bounds().Dx()/2), screenHeight/2)
+	screen.DrawImage(img, op)
+}
+
+func (g *Game) Layout(outsideWidth, outsideHeight int) (int, int) {
+	return screenWidth, screenHeight
+}
+
+func main() {
+	// Decode image from a byte slice instead of a file so that
+	// this example works in any working directory.
+	// If you want to use a file, there are some options:
+	// 1) Use os.Open and pass the file to the image decoder.
+	//    This is a very regular way, but doesn't work on browsers.
+	// 2) Use ebitenutil.OpenFile and pass the file to the image decoder.
+	//    This works even on browsers.
+	// 3) Use ebitenutil.NewImageFromFile to create an ebiten.Image directly from a file.
+	//    This also works on browsers.
+	rawimg, _, err := image.Decode(bytes.NewReader(images.Ebiten_png))
+	if err != nil {
+		log.Fatal(err)
+	}
+	img, _ = ebiten.NewImageFromImage(rawimg, ebiten.FilterDefault)
+
+	ebiten.SetWindowSize(screenWidth*2, screenHeight*2)
+	ebiten.SetWindowTitle("Audio Panning Loop (Ebiten Demo)")
+	g := &Game{}
+	if err := ebiten.RunGame(g); err != nil {
+		log.Fatal(err)
+	}
+}
+
+// StereoPanStream is an audio buffer that changes the stereo channel's signal
+// based on the Panning.
+type StereoPanStream struct {
+	*bytes.Reader
+	pan float64 // -1: left; 0: center; 1: right
+}
+
+const sqrt2div2 = 0.7071067811865476 // math.sqrt(2)/2.0
+const rad45 = math.Pi / 4            // 45º
+
+func (s *StereoPanStream) Read(p []byte) (n int, err error) {
+	n, err = s.Reader.Read(p)
+	if err != nil {
+		return
+	}
+
+	angle := rad45 * s.pan
+	acos := math.Cos(angle)
+	asin := math.Sin(angle)
+	ls := sqrt2div2 * (acos - asin)
+	rs := sqrt2div2 * (acos + asin)
+	for i := 0; i < len(p); i += 4 {
+		lc := int16(float64(int16(p[i])|int16(p[i+1])<<8) * ls)
+		rc := int16(float64(int16(p[i+2])|int16(p[i+3])<<8) * rs)
```

Comment:
Would it be possible to have mixing the left and right channel like Chromium does? (Though ragtime.ogg is a mono-channel audio...)

Reply:
So the implementation would be like this:

```go
panrad := math.Abs(s.pan) * math.Pi / 2
lgain := math.Cos(panrad)
rgain := math.Sin(panrad)
for i := 0; i < len(p); i += 4 {
    l := float64(int16(p[i])|int16(p[i+1])<<8)
    r := float64(int16(p[i+2])|int16(p[i+3])<<8)
    var lc, rc int16
    if (s.pan <= 0) {
        lc = int16(l + r * lgain)
        rc = int16(r * rgain)
    } else {
        lc = int16(l * lgain)
        rc = int16(r + l * rgain)
    }
    ...
}
```

Reply:
I tested this, and when it gets closes to -1 or 1, the audio becomes clipped (some frequencies get clipped to max), so I adjusted to 
```go
if s.pan <= 0 {
	lc = int16((l + r*lgain) / 2)
	rc = int16(r * rgain)
} else {
	lc = int16(l * lgain)
	rc = int16((r + l*rgain) / 2)
}
```

Reply:
It solves clipping, but still doesn't do a proper mixing of `lc` or `rc` (since `lgain` and `rgain` max out at cos/sin of 45º). Another problem is that it reduces the difference between different pans (if pan is <= 0, lc is getting half of the original value regardless of the pan value).

Reply:
Thank you for testing! What about capping the value after the calculation?

```go
    l := float64(int16(p[i])|int16(p[i+1])<<8)
    r := float64(int16(p[i+2])|int16(p[i+3])<<8)
    var lc, rc int16
    if (s.pan <= 0) {
        lc = int16(l + r * lgain)
        rc = int16(r * rgain)
    } else {
        lc = int16(l * lgain)
        rc = int16(r + l * rgain)
    }
    lc = min(max(lc, 1.0), -1.0)
    rc = min(max(rc, 1.0), -1.0)
```

I'd like to keep the general algorithm as possible...

Reply:
The best way to ensure, even if the user imports a heavily mixed true stereo sound for panning, that the sound is mixed would be:
```go
	angle := deg45 * s.pan
	acos := math.Cos(angle)
	asin := math.Sin(angle)
	ls := sqrt2div2 * (acos - asin)
	rs := sqrt2div2 * (acos + asin)
	for i := 0; i < len(p); i += 4 {
		l := float64(int16(p[i]) | int16(p[i+1])<<8)
		r := float64(int16(p[i+2]) | int16(p[i+3])<<8)
		mixed := (l + r) / 2 // we mix left and right to enforce that the source is mono
		lc := int16(mixed * ls)
		rc := int16(mixed * rs)

		p[i] = byte(lc)
		p[i+1] = byte(lc >> 8)
		p[i+2] = byte(rc)
		p[i+3] = byte(rc >> 8)
	}
```
What do you think?

Reply:
Hmm, does this mean all the audio sources are treated as mono-channel? I'd want to keep the source stereo.

BTW, in the Chromium implementation, deg45 does not exist, right? (I think you are referring the implementation in the StackOverflow question, but I'd want to avoid this since I don't understand the algorithm.)

Reply:
By the way, I tested with capping the values to min/max int16 but clipping still occurs:
```go
	pan := s.pan
	if pan <= 0 {
		pan += 1
	}
	panrad := pan * math.Pi / 2
	lgain := math.Cos(panrad)
	rgain := math.Sin(panrad)
	for i := 0; i < len(p); i += 4 {
		l := float64(int16(p[i]) | int16(p[i+1])<<8)
		r := float64(int16(p[i+2]) | int16(p[i+3])<<8)
		var lc, rc int
		if s.pan <= 0 {
			lc = int(l + r*lgain)
			rc = int(r * rgain)
		} else {
			lc = int(l * lgain)
			rc = int(r + l*rgain)
		}
		lc16 := int16(maxint(minint(lc, math.MaxInt16), math.MinInt16))
		rc16 := int16(maxint(minint(rc, math.MaxInt16), math.MinInt16))

		p[i] = byte(lc16)
		p[i+1] = byte(lc16 >> 8)
		p[i+2] = byte(rc16)
		p[i+3] = byte(rc16 >> 8)
	}
```

Just to show that implementations may differ, Unity does this:

> AudioSource.panStereo. 
> Description
> 
> Pans a playing sound in a stereo way (left or right). This only applies to sounds that are Mono or Stereo.
> 
> This pan is applied before 3D panning calculations are considered. In other words, stereo panning affects the left right balance of the sound before it is spatialised in 3D.
> 
> Mono sounds are panned from left to right using constant power panning (non linear fade). This means when pan = 0.0, the balance for the sound in each speaker is 71% left and 71% right, not 50% left and 50% right. This gives (audibly) smoother pans.
> 
> Stereo sounds heave each left/right value faded up and down according to the specified pan position. This means when pan = 0.0, the balance for the sound in each speaker is 100% left and 100% right. When pan = -1.0, only the left channel of the stereo sound is audible, when pan = 1.0, only the right channel of the stereo sound is audible.
> 
> Values range from -1.0 to 1.0.
> 
> -1.0 = Full left 0.0 = center 1.0 = full right
> 
> Default = 0.0.

TLDR: If mono, use 45 degrees formula; If stereo: use my original implementation (use original left and right channels, linear gain instead of 15degrees).

In my opinion, since Ebiten doesn't keep track if the original sound was mono or stereo, I would force into mono and do the 45 degrees formula.

Reply:
Ah ok, referring an existing API sounds great! I'm fine as long as the algorithm is well documented and widely used. Could you leave a comment that the implementation is in the same way of [AudioSource.panStereo](https://docs.unity3d.com/ScriptReference/AudioSource-panStereo.html)? 

> In my opinion, since Ebiten doesn't keep track if the original sound was mono or stereo, I would force into mono and do the 45 degrees formula.

I slightly prefer to always treat the source as stereo, but treating them as mono would make sense as an example. Please also comment about this.

Reply:
If you want, I can revert the algorithm back to the one that uses the original channels (like Unity does when it's stereo). For mono, it would't make much difference (it would only be a little less smooth).

The 45 degrees formula ensures that the `sqrt(lgain*lgain + rgain*rgain)` stays at 1:
![img](https://i.imgur.com/lumZ8Or.png)

Reply:
> If you want, I can revert the algorithm back to the one that uses the original channels (like Unity does when it's stereo). For mono, it would't make much difference (it would only be a little less smooth).

Yes, the stereo way is nicer to me. Sorry for going back and forth.

The unclear thing to me is that how the volumes are interpolated: if the panning is -0.9, would the right channel's volume be 0.1? Or, would sin/cos be used here?

> The 45 degrees formula ensures that the sqrt(lgain*lgain + rgain*rgain) stays at 1:

I understand the total power of the waves is always 1. If the panning is -1, would the degree -90°?

Reply:
The range is always between -45º and 45º, so -90º would never happen (since panning is clamped between -1 and 1).
If the panning is -0.9, the angle would be translated to `-40.5º`:
```
// -40.5 to radians: -0.7068583470577035
// cos(-40.5º) : 0.7604059656000309
// sin(-40.5º): -0.6494480483301837
lgain = sqrt2div2 * (cos - sin) // = 0.996917333733128
rgain = sqrt2div2 * (cos + sin) // = 0.07845909572784494
```
I may have miscalculated the true left and right gains of my explanation image 😅

I'll revert the algorithm back to stereo + linear then!

Reply:
I reverted to stereo+linear algorithm

Reply:
OK thank you for explaining, and reverting the implementation! I'll take a look.

Reply:
Please leave a comment about the algorithm and the link to Unity.

Reply:
added!